jwst/outlier_detection/utils.py

```diff
@@ -30,8 +31,104 @@ def create_cube_median(cube_model, maskpt):
+    """
+    # Compute the weight threshold for each input model
+    weight_thresholds = []
+    with resampled_models:
+        for resampled in resampled_models:
+            weight = resampled.wht
+            weight_threshold = compute_weight_threshold(weight, maskpt)
+            weight_thresholds.append(weight_threshold)
+            # close and delete the model, just to explicitly try to keep the memory as clean as possible
+            resampled_models.shelve(resampled, modify=False)
+
+    # compute median over all models
+    if not on_disk:
+        # easier case: all models in library can be loaded into memory at once
+        model_list = []
+        with resampled_models:
+            for resampled in resampled_models:
+                model_list.append(resampled.data)
+                resampled_models.shelve(resampled, modify=False)
+        return np.nanmedian(np.array(model_list), axis=0)
+    else:
+        # set up buffered access to all input models
+        with resampled_models:
+            example_model = resampled_models.borrow(0)
+            shp = example_model.data.shape
+            dtype = example_model.data.dtype
+            nsections, section_nrows = _compute_buffer_indices(example_model, buffer_size)
+            resampled_models.shelve(example_model, modify=False)
+            del example_model
+
+        # get spatial sections of library and compute timewise median, one by one
+        resampled_sections = _get_sections_library(resampled_models, nsections, section_nrows, shp[0])
+        median_image_empty = np.empty(shp, dtype) * np.nan
+        return _create_median(resampled_sections, resampled_models, weight_thresholds, median_image_empty)
+
+
+def _get_sections_library(library, nsections, section_nrows, imrows):
+    """Iterator to return sections from a ModelLibrary.
+
+    Parameters
+    ----------
+    library : ModelLibrary
+        The input data models.
+
+    nsections : int
+        The number of spatial sections in each model
+
+    section_nrows : int
+        The number of rows in each section
+
+    imrows : int
+        The total number of rows in the image
+    """
+    with library:
+        example_model = library.borrow(0)
+        dtype = example_model.data.dtype
+        dtype_wht = example_model.wht.dtype
+        shp = example_model.data.shape
+        library.shelve(example_model, 0, modify=False)
+        del example_model
+    for i in range(nsections):
+        row1 = i * section_nrows
+        row2 = min(row1 + section_nrows, imrows)
+        
+        data_list = np.empty((len(library), row2 - row1, shp[1]), dtype)
+        weight_list = np.empty((len(library), row2 - row1, shp[1]), dtype_wht)
+        with library:
+            for j, model in enumerate(library):
+                data_list[j] = model.data[row1:row2]
+                weight_list[j] = model.wht[row1:row2]
+                library.shelve(model, j, modify=False)
+                del model
```

Comment:
I think this `del` is unnecessary; that's sort of the point of ModelLibrary

Reply:
It is a minor improvement due to the scope of `model`.
For the first return of `enumerate` `model` will be assigned the first `DataModel`, then the body of the `for` loop run (1x `DataModel` in memory). Without this `del`, the first `DataModel` will remain in memory until the start of the 2nd loop which will (inside the `enumerate`) open a 2nd `DataModel` (so 2x `DataModel`s in memory) before assinging the 2nd `DataModel` to `model` (which will then release the only reference to the first `DataModel`).
tldr; the `del` allows 1x instead of 2x `DataModel`s in memory.
I'm not sure this is worth the `del` since having 2x in memory since the models are relatively small.
See https://github.com/spacetelescope/stpipe/issues/173 for related issue (and linked fix) for `map_function`.

Reply:
See https://github.com/spacetelescope/jwst/pull/8683/files#r1725037911 for a suggestion on where to add this `del` to the updated code. I think it's worth doing to free up ~1 datamodel worth of memory.